File: 114-flatten-binary-tree-to-linked-list/114-flatten-binary-tree-to-linked-list.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    
    public void flatten(TreeNode root) {
        // 1.take temp-left and temp-right to store the left and right child.
        //     2.make node-left=null
        //     3.call flatter(temp-left) and flatten(temp-right)
        //     4.attach the left part to node's right
            // 5. traverse till the bottommost node in the attached part and attach the right to it.
        if(root==null)
            return;
        TreeNode temp_left=root.left;
        TreeNode temp_right=root.right;
        root.left=null;
        flatten(temp_left);
        flatten(temp_right);
        root.right=temp_left;
        TreeNode current_node=root;
        while(current_node.right!=null){
            current_node=current_node.right;
        }
        current_node.right=temp_right;
    }
}

File: 114-flatten-binary-tree-to-linked-list/114-flatten-binary-tree-to-linked-list.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    void flatten(TreeNode* root) {
        // solution
        // 1. first flatten left and right subtree
        //2. store the tail pointers of left and right subtree
        // 3. store the right subtree in a temp var
        // 4. now make left subtree as right subtree and then join the left tail pointer with right subtree
        if(root==NULL ||(root->left==NULL && root->right==NULL)){
            return;
        }
        if(root->left!=NULL){
            //here we flatten the left subtree
            flatten(root->left);
            //now store right subtree into temp
            TreeNode* temp=root->right;
            //now make left subtree as right subtree
            root->right=root->left;
            //makes left as NULL
            root->left=NULL;
            
            //now join tree's tail pointer with right subtree
            //since we know we joined left subtree as right subtree and there is no left subtree
            TreeNode* t=root->right;
            while(t->right!=NULL)
                t=t->right;
            //now joins to right subtee
            t->right=temp;
            
        }
        //flatten the right subtree
        flatten(root->right);
    }
};